.github/workflows/build.yml

```diff
@@ -484,6 +484,56 @@ jobs:
     - name: Tests
       run: xvfb-run make test
 
+  build_tsan:
+    name: 'Thread sanitizer'
+    runs-on: ubuntu-20.04
+    timeout-minutes: 60
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+    - uses: actions/checkout@v4
+    - name: Install Dependencies
+      run: sudo ./.github/workflows/posix-deps-apt.sh
+    - name: Set up GCC-11 for ASAN
+      uses: egor-tensin/setup-gcc@v1
+      with:
+        version: 11
+    - name: TSAN Option Setup
+      run: echo "TSAN_OPTIONS=suppressions=${GITHUB_WORKSPACE}/Tools/tsan/supressions.txt" >> $GITHUB_ENV
+    - name: Configure CPython
+      run: ./configure --with-thread-sanitizer --without-pymalloc --disable-ipv6
+    - name: Build CPython
+      run: make -j4
+    - name: Display build info
+      run: make pythoninfo
+    - name: Tests
+      run: ./python -m test --tsan -j4
+
+  build_tsan_free_threading:
+    name: 'Thread sanitizer (free-threading)'
+    runs-on: ubuntu-20.04
+    timeout-minutes: 60
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+    - uses: actions/checkout@v4
+    - name: Install Dependencies
+      run: sudo ./.github/workflows/posix-deps-apt.sh
+    - name: Set up GCC-11 for ASAN
```

Comment:
Again, you can either use Ubuntu 22.04 (it will give you gcc 12.3), or you can switch to clang which is generally better with sanitizers.

Reply:
Ah got it, I missed what you pointed out, Thanks!

Reply:
Maybe we should try clang as you commented.